packages/@lwc/template-compiler/src/shared/types.ts

```diff
@@ -184,7 +190,25 @@ export interface Slot extends AbstractBaseElement {
     slotName: string;
 }
 
-export type BaseElement = Element | ExternalComponent | Component | Slot;
+// Special LWC tag names denoted with lwc:*
+export interface BaseLwcElement<T extends `${LwcTagName}`> extends AbstractBaseElement {
+    type: 'Lwc';
+    name: T;
+}
+
+/**
+ * Node representing the lwc:component element
+ */
+export interface LwcComponent extends BaseLwcElement<'lwc:component'> {}
+
+/**
+ * All supported special LWC tags, they should all begin with lwc:*
+ */
+export enum LwcTagName {
+    Component = 'lwc:component',
+}
+
+export type BaseElement = Element | ExternalComponent | Component | Slot | LwcComponent;
```

Comment:
I created the AST nodes to look like this because I assumed we would have more `lwc:*` elements in the future.  

I felt that this could help set the basis for future `lwc:*` components but it may also be premature to create the nodes structure ahead of time.  

I can change this to a single AST node if the team prefers.

Reply:
Let's plan for more `lwc:*` nodes. I think it will come in the next release. (`lwc:host`)